[sharded_tensor] simplify init_from_local_shards API

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66351
* __->__ #64481

This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Commit:
[sharded_tensor] simplify init_from_local_shards API

This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

[ghstack-poisoned]

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -222,7 +222,7 @@ def zeros(sharding_spec: ShardingSpec,
 def full(sharding_spec: ShardingSpec,
-         size,
+         *size,
          fill_value=torch.types.Number,
```
nit: unrelated to this PR?

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -289,11 +289,8 @@ def init_from_local_shards(
     Args:
```
update cmt abt global_metadata ?

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -289,11 +289,8 @@ def init_from_local_shards(
-            of the ShardedTensor, must comply with `local_shards` defined in each rank.
-            Note that `sharded_tensor_metadata` must be valid and should also contain
-            local shards metadata.
+        overall_size (int...):  a list, tuple, or `torch.Size` of integers defining the
```
nit: overall_size or global_size ?

bowangbj commented on torch/distributed/_sharding_spec/_internals.py:
```diff
@@ -122,3 +123,15 @@ def check_tensor(shards_metadata, tensor_dims) -> None:
             f'all the individual shards do not cover the entire tensor')
+
+def check_tensor_size(size) -> List[int]:
+    if len(size) == 1 and isinstance(size[0], collections.Sequence):
```
nit: maybe add a cmt and add an example to make it easier to understand?

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
+            local_shard_tensor = local_shard.tensor
+            local_shard_meta = local_shard.metadata
+            local_shard_metadatas.append(local_shard_meta)
+            rank, local_device = sharded_tensor._parse_and_validate_remote_device(local_shard_meta.placement)
```
should we also validate rank is the same as current rank?

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
 
-            if local_shard_tensor.device != local_device:
+        # build a "local" ShardedTensorMetadata with all local shards on this rank
+        global_size = torch.Size(check_tensor_size(overall_size))
```
nit: 

s/check_tensor_size/validate_and_flatten since what is done here is not just "checking" :)

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
 
-            if current_rank == rank:
-                local_shard_metadatas.append(shard_metadata)
+        # First, do local validations on local shards
```
1. Validate local tensors

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
+                pin_memory=local_shards_is_pinned
+            )
+        )
+
```
2. all_gather to collect local_shard_metadata

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
+        # build a "local" ShardedTensorMetadata with all local shards on this rank
+        global_size = torch.Size(check_tensor_size(overall_size))
+
+        local_shard_tensor_metadata = ShardedTensorMetadata(
```
local_sharded_tensor_metadata

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
-        sharded_tensor = cls.__new__(cls)
-
         # prepare initialization
+        sharded_tensor = cls.__new__(cls)
```
nit: here we New ShardedTensor b/f the construct param is ready (eg., sharded tensor metadata). 

does it make sense to do validation first to collect all required params first
once validation and all_gather are done -- we create sharded tensor and init it right way and return?

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
+            )
+        )
+
+        # do all_gather to gather all sharded tensor metadatas in each rank
```
nit: 

2. all_gather to collect local_sharded_tensor_metadata from all ranks

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
+
+        # do all_gather to gather all sharded tensor metadatas in each rank
+        all_metadatas: List[Optional[ShardedTensorMetadata]] = [None for _ in range(world_size)]
+
```
nit: 

s/all_metadatas/gathered_metadatas

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
+            group=sharded_tensor._process_group
+        )
+
+        # do validation with gathered metadatas and build up a global sharded tensor metadata
```
3. Validate the gathered local_sharded_metadata and build up a global sharded_tensor_metadata

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
+
+        # do validation with gathered metadatas and build up a global sharded tensor metadata
+        global_sharded_tensor_metadata = all_metadatas[0]
+
```
nit: should we check size first?

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
+
+        for i in range(1, len(all_metadatas)):
+            global_sharded_tensor_metadata.shards_metadata.extend(all_metadatas[i].shards_metadata)
+            if global_sharded_tensor_metadata.size != all_metadatas[i].size:
```
nit: similarly probably validate first before extending it to global_sharded_tensor_metadata

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -289,11 +289,8 @@ def init_from_local_shards(
-            Note that `sharded_tensor_metadata` must be valid and should also contain
-            local shards metadata.
+        overall_size (int...):  a list, tuple, or `torch.Size` of integers defining the
+            shape of the overall sharded tensor.
```
probably add an example to showcase how to use it?

Example:

rank-0
- local_shards  =...
- global_size = (10, 20)
- sharded_tensor = ShardedTensor.init_from_local(local_shards)

rank-1
- local_shards  =...
- global_size = (10, 20)
- sharded_tensor = ShardedTensor.init_from_local(local_shards)

bowangbj reviewed (commented):
Thanks Wanchao, looks pretty good. Just a few styling issues.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
+        # do all_gather to gather all sharded tensor metadatas in each rank
+        all_metadatas: List[Optional[ShardedTensorMetadata]] = [None for _ in range(world_size)]
+
+        dist.all_gather_object(
```
Don't we need to set the cuda device here appropriately?

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

[ghstack-poisoned]

bowangbj commented:
hey wanchao, saw some update from last review. is this ready for another view? thx

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

wanchaol commented:
> hey wanchao, saw some update from last review. is this ready for another view? thx

@bowangbj @pritamdamania thanks for the comments, this should be ready for another review now :)

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

bowangbj reviewed (commented):
Thanks Wanchao, looks much better and cleaner. Just a few style issues how to condense the code and make it more readable.

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] simplify init_from_local_shards API"


This simplifies `init_from_local_shards` API in sharded tensor, to only require user pass in a list of `Shard` and `overall_size`, instead of ShardedTensorMetadata. We will do the all_gather inside to form a valid ShardedTensorMetadata instead.

TODO: add more test cases to improve coverage.

Differential Revision: [D30748504](https://our.internmc.facebook.com/intern/diff/D30748504/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30748504/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]
